### PR TITLE
created Arm

### PR DESCRIPTION
**Pull request by Spikes-2212**

### CODE REVIEW

File: src/main/java/frc/robot/RobotMap.java [hunk 1]
```diff
@@ -4,6 +4,9 @@ public class RobotMap {
 
     public interface CAN {
 
+        int ARM_FIRST_JOINT_SPARKMAX_1 = -1;
+        int ARM_FIRST_JOINT_SPARKMAX_2 = -1;
+        int ARM_SECOND_JOINT_SPARKMAX_1 = -1;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java]: second joint has only one sparkmax, it doesn't need any indexing

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        return this.trapezoidProfileSettings;
+    }
+
+    public double getEncoderPosition() {
+        return encoder.getPosition();
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putNumber("first joint encoder position", this::getEncoderPosition);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: no need to write first joint. this is already in the first joint namespace

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return instance;
+    }
+
+    private ArmFirstJoint(String namespaceName, CANSparkMax master, CANSparkMax slave) {
+        super(namespaceName, master, slave);
+        this.encoder = master.getAlternateEncoder(DISTANCE_PER_PULSE);
+        slave.follow(master, true);
+        this.PIDSettings = new PIDSettings(kP, kI, kD, waitTime, tolerance);
+        this.feedForwardSettings = new FeedForwardSettings(kS, kV, kA, kG);
+        this.trapezoidProfileSettings = new TrapezoidProfileSettings(trapezoidVelocity, trapezoidAcceleration);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: add a call to configureDashboard in the end of the constructor

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    private final RelativeEncoder encoder;
+
+    private final Namespace PIDNamespace = namespace.addChild("pid");
+    private final Supplier<Double> kP = PIDNamespace.addConstantDouble("kP", 0);
+    private final Supplier<Double> kI = PIDNamespace.addConstantDouble("kI", 0);
+    private final Supplier<Double> kD = PIDNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> waitTime = PIDNamespace.addConstantDouble("wait time", 0);
+    private final Supplier<Double> tolerance = PIDNamespace.addConstantDouble("tolerance", 0);
+    private final com.spikes2212.control.PIDSettings PIDSettings;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: just write PIDSettings, don't write the whole package name

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    private ArmSecondJoint(String namespaceName, CANSparkMax sparkMax) {
+        super(namespaceName, sparkMax);
+        this.encoder = sparkMax.getAlternateEncoder(DISTANCE_PER_PULSE);
+        this.PIDSettings = new PIDSettings(kP, kI, kD, waitTime, tolerance);
+        this.feedForwardSettings = new FeedForwardSettings(kS, kV, kA, kG);
+        this.trapezoidProfileSettings = new TrapezoidProfileSettings(trapezoidVelocity, trapezoidAcceleration);
+    }
+
+    public static ArmSecondJoint getInstance() {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java:51]: move this above the constructor

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        return this.trapezoidProfileSettings;
+    }
+
+    public double getEncoderPosition() {
+        return encoder.getPosition();
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putNumber("second joint encoder position", this::getEncoderPosition);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: no need to write second joint..

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    private final Supplier<Double> trapezoidAcceleration = trapezoidSettingsNamespace.addConstantDouble
+            ("acceleration", 0);
+    private final TrapezoidProfileSettings trapezoidProfileSettings;
+
+    private ArmSecondJoint(String namespaceName, CANSparkMax sparkMax) {
+        super(namespaceName, sparkMax);
+        this.encoder = sparkMax.getAlternateEncoder(DISTANCE_PER_PULSE);
+        this.PIDSettings = new PIDSettings(kP, kI, kD, waitTime, tolerance);
+        this.feedForwardSettings = new FeedForwardSettings(kS, kV, kA, kG);
+        this.trapezoidProfileSettings = new TrapezoidProfileSettings(trapezoidVelocity, trapezoidAcceleration);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: add a call to configureDashboard in the end of the constructor

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        configureDashboard();
+    }
+
+    @Override
+    public void configureLoop(PIDSettings PIDSettings, FeedForwardSettings feedForwardSettings,
+                              TrapezoidProfileSettings trapezoidProfileSettings) {
+        super.configureLoop(PIDSettings, feedForwardSettings, trapezoidProfileSettings);
+        this.setConversionRates();
+    }
+
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java:85]: remove blank line

File: src/main/java/frc/robot/RobotMap.java [hunk 9]
```diff
@@ -4,6 +4,11 @@ public class RobotMap {
 
     public interface CAN {
 
+        int ARM_FIRST_JOINT_SPARKMAX_1 = -1;
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/RobotMap.java]: rename all of these to master and slaves accordingly

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                    new CANSparkMax(RobotMap.CAN.ARM_FIRST_JOINT_SPARKMAX_3, CANSparkMaxLowLevel.MotorType.kBrushless)
+            );
+        }
+        return instance;
+    }
+
+    private ArmFirstJoint(String namespaceName, CANSparkMax master, CANSparkMax slave1, CANSparkMax slave2) {
+        super(namespaceName, master, slave1, slave2);
+        this.encoder = master.getEncoder();
+        encoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: call `setConversionRates()` instead

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+    @Override
+    public void configureLoop(PIDSettings PIDSettings, FeedForwardSettings feedForwardSettings,
+                              TrapezoidProfileSettings trapezoidProfileSettings) {
+        super.configureLoop(PIDSettings, feedForwardSettings, trapezoidProfileSettings);
+        this.setConversionRates();
+    }
+
+
+    public void setConversionRates() {
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: rename to `setConversionFactors()`

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+            );
+            return instance;
+        }
+        return instance;
+    }
+
+    private ArmSecondJoint(String namespaceName, CANSparkMax master, CANSparkMax slave) {
+        super(namespaceName, master, slave);
+        this.encoder = master.getEncoder();
+        encoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: call `setConversionRates()` instead

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        slave.follow(master, false);
```
Comment (OfriRosenbaum) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: no need for this, slave already follows master (look in the super constructor)

**Review by OfriRosenbaum — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/RobotMap.java (hunk 9 above)
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java]: change everything to master/slave

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+public class ArmFirstJoint extends SparkMaxGenericSubsystem {
+
+    public static final int DISTANCE_PER_PULSE = -1;
+
+    public static final int SECONDS_IN_MINUTE = 60;
+
+    private static ArmFirstJoint instance;
+
+    private final RelativeEncoder encoder;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: add `TODO check encoder type`

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+    public FeedForwardSettings getFeedForwardSettings() {
+        return this.feedForwardSettings;
+    }
+
+    public TrapezoidProfileSettings getTrapezoidProfileSettings() {
+        return this.trapezoidProfileSettings;
+    }
+
+    public double getEncoderPosition() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: A-III\a6

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+        return encoder.getPosition();
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java:115]: ditto

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+        encoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
+        slave2.follow(master, true);
+        this.PIDSettings = new PIDSettings(kP, kI, kD, waitTime, tolerance);
+        this.feedForwardSettings = new FeedForwardSettings(kS, kV, kA, kG);
+        this.trapezoidProfileSettings = new TrapezoidProfileSettings(trapezoidVelocity, trapezoidAcceleration);
+        configureDashboard();
+    }
+
+    @Override
+    public void configureLoop(PIDSettings PIDSettings, FeedForwardSettings feedForwardSettings,
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: How tf does this code compile with two `PIDSettings`

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+
+    public static final int DISTANCE_PER_PULSE = -1;
+
+    public static final int SECONDS_IN_MINUTE = 60;
+
+    private static ArmSecondJoint instance;
+
+    private final RelativeEncoder encoder;
+
+    private final Namespace PIDNamespace = namespace.addChild("pid");
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: A-I\a1

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+
+public class ArmSecondJoint extends SparkMaxGenericSubsystem {
+
+    public static final int DISTANCE_PER_PULSE = -1;
+
+    public static final int SECONDS_IN_MINUTE = 60;
+
+    private static ArmSecondJoint instance;
+
+    private final RelativeEncoder encoder;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: move below constants

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+                    new CANSparkMax(RobotMap.CAN.ARM_FIRST_JOINT_SPARKMAX_2, CANSparkMaxLowLevel.MotorType.kBrushless),
+                    new CANSparkMax(RobotMap.CAN.ARM_FIRST_JOINT_SPARKMAX_3, CANSparkMaxLowLevel.MotorType.kBrushless)
+            );
+        }
+        return instance;
+    }
+
+    private ArmFirstJoint(String namespaceName, CANSparkMax master, CANSparkMax slave1, CANSparkMax slave2) {
+        super(namespaceName, master, slave1, slave2);
+        this.encoder = master.getEncoder();
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: alternate/absolute encoder

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 21]
```diff
@@ -0,0 +1,102 @@
...
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.control.TrapezoidProfileSettings;
+import com.spikes2212.dashboard.Namespace;
+import frc.robot.RobotMap;
+
+import java.beans.Encoder;
+import java.util.function.Supplier;
+
+public class ArmFirstJoint extends SparkMaxGenericSubsystem {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java:16]: Bad name

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.control.TrapezoidProfileSettings;
+import com.spikes2212.dashboard.Namespace;
+import frc.robot.RobotMap;
+
+import java.beans.Encoder;
+import java.util.function.Supplier;
+
+public class ArmSecondJoint extends SparkMaxGenericSubsystem {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java:16]: bad name

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 23]
```diff
@@ -0,0 +1,103 @@
...
+
+    private final RelativeEncoder encoder;
+
+    private final Namespace PIDNamespace = namespace.addChild("pid");
+    private final Supplier<Double> kP = PIDNamespace.addConstantDouble("kP", 0);
+    private final Supplier<Double> kI = PIDNamespace.addConstantDouble("kI", 0);
+    private final Supplier<Double> kD = PIDNamespace.addConstantDouble("kD", 0);
+    private final Supplier<Double> waitTime = PIDNamespace.addConstantDouble("wait time", 0);
+    private final Supplier<Double> tolerance = PIDNamespace.addConstantDouble("tolerance", 0);
+    private final PIDSettings PIDSettings;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: ditto

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+
+    public FeedForwardSettings getFeedForwardSettings() {
+        return this.feedForwardSettings;
+    }
+
+    public TrapezoidProfileSettings getTrapezoidProfileSettings() {
+        return this.trapezoidProfileSettings;
+    }
+
+    public double getEncoderPosition() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: A-III\a6

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 25]
```diff
@@ -0,0 +1,103 @@
...
+        return encoder.getPosition();
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java:113]: ditto

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java (hunk 24 above)
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: rename to `getPosition`

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/MoveArm.java [hunk 26]
```diff
@@ -0,0 +1,32 @@
...
+import com.spikes2212.command.genericsubsystem.commands.smartmotorcontrollergenericsubsystem.MoveSmartMotorControllerSubsystemTrapezically;
+import edu.wpi.first.wpilibj2.command.ParallelCommandGroup;
+import frc.robot.subsystems.ArmFirstJoint;
+import frc.robot.subsystems.ArmSecondJoint;
+
+public class MoveArm extends ParallelCommandGroup {
+
+    public enum ArmState {
+
+        REST(0, 0), FLOOR(0, 0), DOUBLE_SUBSTATION(0, 0), BOTTOM(0, 0), CUBE_MID(0, 0),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/MoveArm.java]: this will be fun to calibrate

File: src/main/java/frc/robot/commands/MoveArm.java [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+            this.firstJointPosition = firstJointPosition;
+            this.secondJointPosition = secondJointPosition;
+        }
+    }
+
+    public MoveArm(ArmFirstJoint firstJoint, ArmSecondJoint secondJoint, ArmState state) {
+        super(new MoveSmartMotorControllerSubsystemTrapezically(firstJoint,
+                        firstJoint.getPIDSettings(), firstJoint.getFeedForwardSettings(),
+                        () -> state.firstJointPosition, firstJoint.getTrapezoidProfileSettings()),
+                new MoveSmartMotorControllerSubsystemTrapezically(secondJoint,
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/MoveArm.java:28]: remind me to talk to you about it, I worry that trapezim won't work

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+        }
+        return instance;
+    }
+
+    private ArmFirstJoint(String namespaceName, CANSparkMax master, CANSparkMax slave) {
+        super(namespaceName, master, slave);
+        sparkMaxEncoder = master.getEncoder();
+        absoluteEncoder = new DutyCycleEncoder(RobotMap.DIO.ARM_FIRST_JOINT_ABSOLUTE_ENCODER);
+        configureEncoders();
+        slave.follow(master, true);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java:68]: unnecessary

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public void configureLoop(PIDSettings pidSettings, FeedForwardSettings feedForwardSettings,
+                              TrapezoidProfileSettings trapezoidProfileSettings) {
+        super.configureLoop(pidSettings, feedForwardSettings, trapezoidProfileSettings);
+        configureEncoders();
+    }
+
+    public double getPosition() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: `getRelativePosition`

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java (hunk 28 above)
Comment (Spikes-2212) [reply src/main/java/frc/robot/subsystems/ArmFirstJoint.java:68]: No

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    public TrapezoidProfileSettings getTrapezoidProfileSettings() {
+        return trapezoidProfileSettings;
+    }
+
+    private void configureEncoders() {
+        sparkMaxEncoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
+        sparkMaxEncoder.setVelocityConversionFactor(DISTANCE_PER_PULSE / SECONDS_IN_MINUTE);
+        absoluteEncoder.setDistancePerRotation(DISTANCE_PER_PULSE);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java]: image

File: src/main/java/frc/robot/subsystems/ArmFirstJoint.java [hunk 31]
```diff
@@ -0,0 +1,114 @@
...
+        sparkMaxEncoder.setPosition(absoluteEncoder.getDistance());
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmFirstJoint.java:115]: add speed

File: src/main/java/frc/robot/RobotMap.java [hunk 32]
```diff
@@ -4,18 +4,23 @@ public class RobotMap {
 
     public interface CAN {
 
+        int ARM_FIRST_JOINT_SPARKMAX_MASTER = -1;
+        int ARM_FIRST_JOINT_SPARKMAX_SLAVE = -1;
+        int ARM_SECOND_JOINT_SPARKMAX_MASTER = -1;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/RobotMap.java:14]: why master if it's the only one?

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 33]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public TrapezoidProfileSettings getTrapezoidProfileSettings() {
+        return trapezoidProfileSettings;
+    }
+
+    private void configureEncoders() {
+        sparkMaxEncoder.setPositionConversionFactor(DISTANCE_PER_PULSE);
+        sparkMaxEncoder.setVelocityConversionFactor(DISTANCE_PER_PULSE / SECONDS_IN_MINUTE);
+        absoluteEncoder.setDistancePerRotation(DISTANCE_PER_PULSE);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java]: Please refer to above meme

File: src/main/java/frc/robot/subsystems/ArmSecondJoint.java [hunk 34]
```diff
@@ -0,0 +1,112 @@
...
+        sparkMaxEncoder.setPosition(absoluteEncoder.getDistance());
+    }
+
+    @Override
+    public void configureDashboard() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/ArmSecondJoint.java:113]: add speed

**Review by Yoel-Perman-Brilliant — CHANGES_REQUESTED** [review]

**Pull request merged**